Use tube identifiers to plate samples

ElDeveloper opened the pull request:
Note, this PR is currently incomplete and still needs a fair amount of work
(mainly surrounding the tests). I've opened this PR to run the full test suite.

ElDeveloper commented on labman/gui/static/js/plateViewer.js:
```diff
@@ -451,7 +467,6 @@ PlateViewer.prototype.updateDuplicates = function () {
       var row = elem[0] - 1;
       var col = elem[1] - 1;
       that.wellClasses[row][col].push('well-duplicated');
-      that.data[row][col] = elem[2];
```
As far as I can tell, this update is not ncessary since this check is should only update the "styling" of the cells, not the content.

ElDeveloper commented:
This is now ready for review. I would appreciate a careful review, specifically of the sample plating interface since this is the main aspect that was changed. The main change is that when a study has a "specimen_id_column" set, then the user can plate using the values in that column. When this column is not set, the user has to plate the samples using the sample identifiers as defined by Qiita. See this example for the same study except one has a specimen_id_column and one doesn't:

## Without specimen_id_column (users need to type the qiita id):
![screen shot 2018-09-06 at 5 43 37 pm](https://user-images.githubusercontent.com/375307/45192315-91281c00-b1fc-11e8-9f5e-ebffbd514798.png)

## With a specimen_id_column (users need to type the specimen identifiers):
![screen shot 2018-09-06 at 5 33 16 pm](https://user-images.githubusercontent.com/375307/45192322-9ab18400-b1fc-11e8-894e-cb6b54c723a1.png)


Here's a few notes:
- I've updated the Travis build to use the current dev branch of Qiita. Since the master branch didn't have the most recent changes that are needed for this PR. This can be changed at a later time.
- The PATCH request that dealt with the well updates will now handle using the specimen identifiers if such a column is present in the study's description.
- Since I cannot modify the specimen_id_column attribute via the Python API (because labman doesn't have access to that attribute), I've decided to instead make direct queries against the DB.
- Should we allow users to plate samples even if no study has been selected? I am not sure why this is possible, but it is. I would be happy to remove this altogether if we can all agree on that.

wasade commented on labman/db/study.py:
```diff
@@ -64,6 +64,120 @@ def creator(self):
+        Raises
+        ------
+        ValueError
+            If not matches are found (when a specimen_id_column is set).
```
not -> no

wasade commented on labman/db/study.py:
```diff
@@ -64,6 +64,120 @@ def creator(self):
+        Notes
+        -----
+        If a specimen identifier column hasn't been set, this method will
+        return the inputed value.
```
inputed -> input

wasade commented on labman/db/tests/test_study.py:
```diff
@@ -82,6 +83,107 @@ def test_samples(self):
+            # case insensitive
+            self.assertEqual(s.samples('Skm'), exp_samples)
+            self.assertEqual(s.samples('64'), [])
+
```
Is there a test of a `limit` but no `term`?

ElDeveloper replied:
Yes there's a test in line 111.

wasade replied:
oh, sorry!

ElDeveloper replied:
👍

wasade commented on labman/db/tests/test_study.py:
```diff
@@ -82,6 +83,107 @@ def test_samples(self):
+
+            # should be an exact match
+            with self.assertRaisesRegex(ValueError,
+                                        'Could not find \"1\.skm4\.640180\"'):
```
Just to check, is case insensitivity a safe property of `sample_name`?

ElDeveloper replied:
Yes I think its fair to require exact case matching.

ElDeveloper commented on labman/db/composition.py:
```diff
@@ -537,6 +537,25 @@ def content(self):
+        if study is not None:
+            return study.sample_id_to_specimen_id(self.sample_id)
+        else:
+            return self.content
```
Yes

AmandaBirmingham commented on labman/db/study.py:
```diff
@@ -77,25 +191,31 @@ def samples(self, term=None, limit=None):
+            if specimen_id_column is None:
+                sql = """SELECT sample_id
+                         FROM qiita.study_sample
+                         WHERE study_id = %s AND LOWER(sample_id) LIKE %s
```
Why is it preferable to get sample_id from qiita.study_sample rather than getting it from the qiita.sample_{studyid} table as is done if there is a specimen_id_column?  Is it better performing or something?  Just curious because I would think that it would be possible (although maybe not advisable? :) to use the current SQL for when there is specimen id in BOTH these cases by doing the below before formatting the sql ...

     if specimen_id_column is None:
            specimen_id_column = "sample_id"

ElDeveloper replied:
This is a great suggestion!

AmandaBirmingham commented on labman/db/tests/test_composition.py:
```diff
@@ -115,6 +116,28 @@ def test_sample_composition_get_control_samples(self):
+        obs = SampleComposition(1).specimen_id
+        # returns the underlying id if no specimen_id_column is set
+        self.assertEqual(obs, '1.SKB1.640202')
+
```
Might be nice to have the sort of blank test that is done after setting a specimen_id (copied below) also done here where one isn't set, to make sure we don't do anything later to screw up the handling of that case :)

            obs = SampleComposition(8).specimen_id
            self.assertEqual(obs, 'blank.21.H1')

ElDeveloper replied:
Great suggestion.

AmandaBirmingham commented on labman/gui/handlers/process_handlers/test/test_sample_plating_process.py:
```diff
@@ -100,12 +108,77 @@ def test_sample_plating_process_handler_patch_request(self):
+        with sql_connection.TRN as TRN:
+            TRN.add(sql, ['anonymized_name'])
+
+            print(Study(1).specimen_id_column)
```
What is the print for?

ElDeveloper replied:
My bad! That has been removed now.

AmandaBirmingham commented on labman/gui/handlers/process_handlers/test/test_sample_plating_process.py:
```diff
@@ -100,12 +108,77 @@ def test_sample_plating_process_handler_patch_request(self):
         self.assertIsNone(tester.notes)
 
+    def test_patch_with_specimen_id(self):
+        # Test success
```
I'm confused by this comment ... the tests right below this comment are just showing the well is what we expect it to be when we start the test, right?  Maybe this comment could be expanded a little bit to make it more obvious?

AmandaBirmingham commented on labman/gui/static/js/plateViewer.js:
```diff
@@ -352,6 +352,20 @@ PlateViewer.prototype.loadPlateLayout = function () {
+  studyID = get_active_studies().pop();
+
+  if (studyID === undefined) {
+    alert('Please select a study, plating samples without selecting a study ' +
```
study, -> study;

ElDeveloper commented:
Thanks for the review @AmandaBirmingham!

> I don't think I understand how the functionality works when samples from more than one study are being plated on the same plate. How does the user switch between studies to pick their preferred active one?

This functionality was already provided with the current interface, so it didn't require any additional changes. For example see in this animated gif how I have two studies to plate both of which have sample names with the letter `s`. However, depending on the study that is selected the suggestions and validations will pop up as expected. At the end of the animation, I add and select another study and type in the letter `a` which reveals only sample names for that third study.

![study-selection](https://user-images.githubusercontent.com/375307/45443913-e770e100-b67a-11e8-9712-e87a2314351b.gif)

Let me know if you would rather go through this in person, I can come find you by your office. In the meantime I'll finish addressing your review comments.

ElDeveloper commented on labman/db/study.py:
```diff
@@ -64,6 +64,120 @@ def creator(self):
+            # if a specimen_id_column is not unique (since this is softly
+            # enforced), then there can be more than one match
+            elif len(res) > 1:
+                raise RuntimeError('More than one match was found, there is a '
```
Good suggestion.

wasade commented:
Awesome!

AmandaBirmingham commented:
Agreed!  Thanks, @ElDeveloper!
